### PR TITLE
# 155514749 enable validation

### PR DESCRIPTION
**Pull request by Ashaba**
#### What does this PR do?
It enables authorization to different pages
#### Description of Task to be completed?
This PR ensures only authorized users can access certain resources.
It implements google OAuth both on the localhost as well as server.
Enables users to be redirected to the home page to login on attempt to access resources unauthenticated
#### How should this be manually tested?
Run `pytest`
Run the server at `localhost:5000/dashboard`
#### Any background context you want to provide?
Previously, pages were open to any user to access yet we needed to track the currently logged in user for actions performed. This feature ensures we have right users access right resources
#### What are the relevant pivotal tracker stories?
155514749

### CODE REVIEW

File: application/auth/helpers.py [hunk 1]
```diff
@@ -28,31 +28,29 @@
 SERVER_ERROR_MSG = 'Authorization failed. Please contact support.'
 
 
-def save_user():
-    user = User.query.filter_by(email=session.get("email")).first()
+def save_user(payload):
+    user = User.query.filter_by(email=payload.get("email")).first()
```
Comment (muhallan) [inline application/auth/helpers.py:32]: Aren't you supposed to validate if email exists?

File: application/auth/helpers.py [hunk 2]
```diff
@@ -28,31 +28,29 @@
...
     if not user:
         user = User(
-            name=session.get("name"),
-            email=session.get("email"),
-            image_url=session.get("image_url")
+            name=payload.get("fullName"),
```
Comment (muhallan) [inline application/auth/helpers.py:35]: I feel that these fields need to first be validated

File: application/auth/helpers.py [hunk 3]
```diff
@@ -28,31 +28,29 @@
 SERVER_ERROR_MSG = 'Authorization failed. Please contact support.'
 
 
-def save_user():
-    user = User.query.filter_by(email=session.get("email")).first()
+def save_user(payload):
+    user = User.query.filter_by(email=payload.get("email")).first()
     if not user:
```
Comment (muhallan) [inline application/auth/helpers.py:33]: What response do you give when the user is not there?

File: application/auth/helpers.py [hunk 4]
```diff
@@ -28,31 +28,29 @@
...
-            if user.name != session.get("name"):
-                setattr(user, 'name', session.get("name"))
-            user.save()
-        save_user()
-        return f(*args, **kwargs)
-    return decorated
+def update_user(payload):
+    user = User.query.filter_by(email=payload.get("email")).first()
+    if user:
+        if user.image_url != payload.get("imageUrl"):
```
Comment (muhallan) [inline application/auth/helpers.py:46]: What if this is blank?

File: application/auth/helpers.py [hunk 5]
```diff
@@ -28,31 +28,29 @@
...
+            setattr(user, 'image_url', payload.get("imageUrl"))
+        if user.name != payload.get("fullName"):
+            setattr(user, 'name', session.get("fullName"))
+        user.save()
+        return user
+    else:
+        save_user(payload)
```
Comment (muhallan) [inline application/auth/helpers.py:53]: You might save a user with an empty payload

File: application/auth/helpers.py [hunk 6]
```diff
@@ -28,31 +28,29 @@
...
         user = User(
-            name=session.get("name"),
-            email=session.get("email"),
-            image_url=session.get("image_url")
+            name=payload.get("fullName"),
+            email=payload.get("email"),
+            image_url=payload.get("imageUrl")
         )
         user.save()
     return user
```
Comment (muhallan) [inline application/auth/helpers.py:40]: What will be returned if the user is not found?

File: application/static/favicon.ico/browserconfig.xml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="utf-8"?>
+<browserconfig><msapplication><tile><square70x70logo src="/ms-icon-70x70.png"/><square150x150logo src="/ms-icon-150x150.png"/><square310x310logo src="/ms-icon-310x310.png"/><TileColor>#ffffff</TileColor></tile></msapplication></browserconfig>
```
Comment (muhallan) [inline application/static/favicon.ico/browserconfig.xml:2]: Need for a new line at the end of the file

File: application/auth/views.py [hunk 8]
```diff
@@ -1,19 +1,34 @@
...
 
 
 @auth.route('/auth', methods=['POST'])
-@token_required
-@update_user
 def authenticate():
-    return jsonify(dict(status="success", message="login successful"))
+    try:
+        payload = request.get_json()
+        update_user(payload)
```
Comment (muhallan) [inline application/auth/views.py:19]: Doesn't this payload need to be checked before being used?

File: application/auth/helpers.py (hunk 1 above)
Comment (Ashaba) [reply application/auth/helpers.py:32]: you mean in the payload? There's no need for that validation as we are using google main, by default that payload contains the user's email

File: application/auth/helpers.py (hunk 2 above)
Comment (Ashaba) [reply application/auth/helpers.py:35]: By default a google user has all those fields so no need to re-validate

File: application/auth/helpers.py (hunk 3 above)
Comment (Ashaba) [reply application/auth/helpers.py:33]: What do you mean? At this stage, I save the user to the db if they don't already exist in the database

File: application/auth/views.py (hunk 8 above)
Comment (Ashaba) [reply application/auth/views.py:19]: Not really, by the time you access the payload, it already has user data

**Comment by eric-elem** [comment]
LGTM

**Comment by muhallan** [comment]
LGTM

**Pull request merged**